### PR TITLE
[receiver/filelog] Implement specifying top n files to track when ordering

### PR DESCRIPTION
**Pull request by BinaryFissionGames**
**Description:**
* Add a new `ordering_criteria.top_n` option, which allows a user to specify the number of files to track after ordering.
  * Default is 1, which was the existing behavior.

**Link to tracking Issue:** #23788

**Testing:**
Unit tests added.

**Documentation:**
Added new parameter to existing documentation.

### CODE REVIEW

File: pkg/stanza/fileconsumer/matcher/matcher.go [hunk 1]
```diff
@@ -62,6 +67,14 @@ func New(c Criteria) (*Matcher, error) {
 		return nil, fmt.Errorf("'regex' must be specified when 'sort_by' is specified")
 	}
 
+	if c.OrderingCriteria.TopN < 0 {
+		return nil, fmt.Errorf("`top_n` must be a positive integer")
```
Comment (djaglowski) [inline pkg/stanza/fileconsumer/matcher/matcher.go:71]:
```suggestion
		return nil, fmt.Errorf("'top_n' must be a positive integer")
```

File: pkg/stanza/fileconsumer/matcher/matcher.go [hunk 2]
```diff
@@ -127,7 +142,10 @@ func (m Matcher) MatchFiles() ([]string, error) {
...
 
-	// Return only the first item.
-	// See https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/23788
-	return result[:1], errors.Join(err, errs)
+	numFiles := m.topN
+	if len(result) < numFiles {
+		numFiles = len(result)
+	}
+
+	return result[:numFiles], errors.Join(err, errs)
```
Comment (djaglowski) [inline pkg/stanza/fileconsumer/matcher/matcher.go:150]: Save the allocation?
```suggestion
	if len(result) <= m.topN {
		return result, errors.Join(err, errs)	
	}
	return result[:m.topN], errors.Join(err, errs)
```

File: pkg/stanza/fileconsumer/matcher/matcher_test.go [hunk 3]
```diff
@@ -98,6 +98,23 @@ func TestNew(t *testing.T) {
...
+					TopN:  -1,
+					SortBy: []Sort{
+						{
+							SortType: "numeric",
+							RegexKey: "key",
+						},
+					},
+				},
+			},
+			expectedErr: "`top_n` must be a positive integer",
```
Comment (djaglowski) [inline pkg/stanza/fileconsumer/matcher/matcher_test.go:116]:
```suggestion
			expectedErr: "'top_n' must be a positive integer",
```

**Review by djaglowski — APPROVED** [review]
LGTM. Just a couple nits

File: pkg/stanza/fileconsumer/matcher/matcher.go (hunk 2 above)
Comment (BinaryFissionGames) [reply pkg/stanza/fileconsumer/matcher/matcher.go]: Done; Also realized I was missing a unit test for the `len(result) <= m.topN ` case, so added that in as well.

File: pkg/stanza/fileconsumer/matcher/matcher.go (hunk 1 above)
Comment (BinaryFissionGames) [reply pkg/stanza/fileconsumer/matcher/matcher.go]: changed this + the test.

**Review by djaglowski — APPROVED** [review]

**Pull request merged**